data/transactions/logic/assembler_test.go

```diff
@@ -2333,6 +2341,21 @@ func TestDigAsm(t *testing.T) {
 
 }
 
+func TestBuryAsm(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	testProg(t, "int 1; bury; +", AssemblerMaxVersion, Expect{1, "bury expects 1 immediate..."})
+	testProg(t, "int 1; bury junk; +", AssemblerMaxVersion, Expect{1, "bury unable to parse..."})
+
+	testProg(t, "int 1; byte 0x1234; int 2; bury 1; +", AssemblerMaxVersion) // the 2 replaces the byte string
+	testProg(t, "int 2; int 2; byte 0x1234; bury 1; +", AssemblerMaxVersion,
+		Expect{1, "+ arg 1..."})
+	testProg(t, "byte 0x32; byte 0x1234; int 2; bury 3; +", AssemblerMaxVersion,
+		Expect{1, "bury 3 expects 4..."})
+	testProg(t, "int 1; byte 0x1234; int 2; bury 12; +", AssemblerMaxVersion,
+		Expect{1, "bury 12 expects 13..."})
```

Comment:
When I traced through this test case I got to [this point of execution](https://github.com/jannotti/go-algorand/blob/20eccb20cf01449461a7a6009a427893171911a7/data/transactions/logic/assembler.go#L1651) with resulting message `"bury 12 expects 13 stack arguments but stack height is 3"`

In general, I would expect it to be impossible to know the stack height at assembly time. It appears that a clever algorithm is provided using `logic.ProgramKnowledge` to catch errors that must always occur. Can you briefly explain how this works? Is it clear that when such an error is flagged there is no way for the stack to have been different at run-time? For example, in this case why do we know that if we were to run a program whose assembler reported this error that its height would be exactly 3 at this point of execution?

Reply:
It is not possible _in general_.  But it is possible sometimes.  The assembler tries to maintain conservative information about the stack (and scratch slots).  It knows some stuff about the stack within a basic block. That is, if operations have occurred that manipulate the stack, and no label is between those ops and the current op, it knows a little bit about the stack.

It's _especially_ knowledgable in the first basic block of the program, because it knows the height of the stack.  In any other block, it can't know, as it could have been jumped to from multiple different places, some of which it hasn't even seen yet.

It also tracks whether it believes the instructions are currently in "dead  code", but that's a different story.

Reply:
So I guess I would gain greater understanding if a follow-up _non-erroring_ test case would be added with the same suffix "int 1; byte 0x1234; int 2; bury 12; +" but some simple instructions (like a label?) that shows the assembler is just fine with this in the right circumstance.

Reply:
prefix `b hello:; hello:;` seems to work

Reply:
Wouldn't any working `bury` use meet this ask?  Like the one a few lines up, or the usage in TestBury of eval_test.go.

Reply:
true. your call. resolving